src/fs/file_system.rs

```diff
@@ -0,0 +1,289 @@
+    }
+
+    /// Tests if the underlying file exists. If this returns `Ok`, the file exists.
+    pub fn try_exists(&mut self, path: &str) -> FileSystemResult<()> {
+        self.metadata(&path).map(|_| ())
+    }
+
+    /// Copies the contents of one file to another. Creates the destination file if it doesn't
+    /// exist and overwrites any content, if it exists.
+    pub fn copy(&mut self, src_path: &str, dest_path: &str) -> FileSystemResult<()> {
+        let read = self.read(src_path)?;
+        self.write(dest_path, read)
+    }
+
+    /// Creates a new, empty directory at the provided path
+    pub fn create_dir(&mut self, path: &str) -> FileSystemResult<()> {
+        let _path = Path::new(path)?;
+        todo!()
+    }
+
+    /// Recursively create a directory and all of its parent components if they are missing.
+    pub fn create_dir_all(&mut self, path: &str) -> FileSystemResult<()> {
+        let _path = Path::new(path)?;
+        todo!()
+    }
+
+    /// Given a path, query the file system to get information about a file, directory, etc.
+    pub fn metadata(&mut self, path: &str) -> FileSystemResult<BoxedUefiFileInfo> {
+        let path = Path::new(path)?;
+        self.resolve_directory_entry_fileinfo(&path)
+    }
+
+    /// Read the entire contents of a file into a bytes vector.
+    pub fn read(&mut self, path: &str) -> FileSystemResult<Vec<u8>> {
+        let path = Path::new(path)?;
+
+        let mut file = self
+            .open_in_root(&path, UefiFileMode::Read)?
+            .into_regular_file()
+            .ok_or(FileSystemError::NotAFile(path.as_str().to_string()))?;
+        let info = self.resolve_directory_entry_fileinfo(&path).unwrap();
+
+        let mut vec = Vec::with_capacity(info.file_size() as usize);
+        vec.resize(vec.capacity(), 0);
+        let read_bytes = file.read(vec.as_mut_slice()).map_err(|e| {
+            log::error!("reading failed: {e:?}");
+            FileSystemError::ReadFailure
+        })?;
+
+        // we read the whole file at once!
+        if read_bytes != info.file_size() as usize {
+            log::error!("Did only read {}/{} bytes", info.file_size(), read_bytes);
+        }
+
+        Ok(vec)
+    }
+
+    /// Returns an iterator over the entries within a directory.
+    pub fn read_dir(&mut self, path: &str) -> FileSystemResult<UefiDirectoryEntryIterator> {
+        let path = Path::new(path)?;
+        let parent_dir = self.open_parent_dir_handle_of_path(&path, UefiFileMode::Read)?;
+        Ok(UefiDirectoryEntryIterator::new(parent_dir))
+    }
+
+    /// Read the entire contents of a file into a string.
+    pub fn read_to_string(&mut self, path: &str) -> FileSystemResult<String> {
+        String::from_utf8(self.read(path)?).map_err(|x| FileSystemError::Utf8Error(x))
+    }
+
+    /// Removes an empty directory.
+    pub fn remove_dir(&mut self, path: &str) -> FileSystemResult<()> {
+        let _path = Path::new(path)?;
+        todo!()
```

Comment:
I've noticed these `todo!` calls. Are you planning to implement them before finalizing the PR, or leave them for future implementation post-merge? (if it's the first case, then I'd advise marking the PR as a draft so we don't accidentally merge it before it's ready 😅)

Reply:
They will be implemented, for sure! I just wanted to clarify if we want this thing at all.

Reply:
Having recently done something with the "raw" filesystem API I can confirm that a simpler fs interface would be nice :)